### PR TITLE
[v7r2] PoolComputingElement updates

### PR DESCRIPTION
**Pull request by fstagni**
With these changes the JobAgent will wait that all processes are finished before finalizing.

BEGINRELEASENOTES

*Reources
CHANGE: PoolComputingElement use concurrent.futures.ProcessPoolExecutor
CHANGE: PoolComputingElemnet: added a shutdown (called by JobAgent)

ENDRELEASENOTES

### CODE REVIEW

**Review by chrisburr — APPROVED** [review]

File: src/DIRAC/Resources/Computing/InProcessComputingElement.py [hunk 1]
```diff
@@ -99,7 +99,7 @@ def submitJob(self, executableFile, proxy=None, **kwargs):
       else:
         error = 'InProcess Job Execution Failed'
       res = S_ERROR(error)
-      res['Value'] = result['Value'][0] - 256
+      res['Value'] = result['Value'][0] - 256  # yes, it's "correct"
```
Comment (aldbr) [inline src/DIRAC/Resources/Computing/InProcessComputingElement.py:102]: Just a minor detail, shouldn't we move this line such as:

```python
elif result['Value'][0] > 128:
  # negative exit values are returned as 256 - exit
  res['Value'] = result['Value'][0] - 256  # yes, it's "correct"
  self.log.warn('InProcess Job Execution Failed')
  self.log.info('Exit status:', result['Value'])
  if result['Value'] == -2:
    error = 'JobWrapper initialization error'
  elif result['Value'] == -1:
    error = 'JobWrapper execution error'
  else:
    error = 'InProcess Job Execution Failed'
  res = S_ERROR(error)
  return res
```

File: src/DIRAC/Resources/Computing/PoolComputingElement.py [hunk 2]
```diff
@@ -263,3 +251,7 @@ def getDescription(self):
     ceDictList.append(dict(ceDict))
 
     return S_OK(ceDictList)
+
+  def shutdown(self):
+    self.pPool.shutdown()
```
Comment (aldbr) [inline src/DIRAC/Resources/Computing/PoolComputingElement.py:256]: From what I understand, the `future.shutdown` will wait until all the processes are done before returning a value.
Maybe you could add a comment about this?

File: src/DIRAC/WorkloadManagementSystem/Agent/JobAgent.py [hunk 3]
```diff
@@ -729,6 +729,12 @@ def finalize(self):
     """ Job Agent finalization method
     """
 
+    res = self.computingElement.shutdown()
```
Comment (aldbr) [inline src/DIRAC/WorkloadManagementSystem/Agent/JobAgent.py:732]: And maybe another comment here to say that `shutdown()` will force the `JobAgent` to wait until all the tasks are in a final state?

File: src/DIRAC/Resources/Computing/InProcessComputingElement.py [hunk 4]
```diff
@@ -89,17 +89,18 @@ def submitJob(self, executableFile, proxy=None, **kwargs):
...
       self.log.error('Fail to run InProcess', result['Message'])
     elif result['Value'][0] > 128:
+      res = S_ERROR()
       # negative exit values are returned as 256 - exit
+      res['Value'] = result['Value'][0] - 256  # yes, it's "correct"
       self.log.warn('InProcess Job Execution Failed')
-      self.log.info('Exit status:', result['Value'][0] - 256)
-      if result['Value'][0] - 256 == -2:
+      self.log.info('Exit status:', result['Value'])
+      if result['Value'] == -2:
```
Comment (fstagni) [inline src/DIRAC/Resources/Computing/InProcessComputingElement.py:97]: @aldbr I followed your advice but here shouldn't it be now
```suggestion
      self.log.info('Exit status:', res['Value'])
      if res['Value'] == -2:
```
?

Comment (aldbr) [reply src/DIRAC/Resources/Computing/InProcessComputingElement.py:97]: Yes you're right, sorry

**Pull request merged**